### PR TITLE
Sequential Timing : setup and hold search

### LINKED ISSUE #61: Flip flop characterization problems.
**Problem Description**
I am trying to characterize a simple flip-flop from a commercial PDK.
It is not a scan flip flop. It has no set / reset asynchronous pins.
I see two problems:
1. The same set of "slews" is used for delay as for "setup/hold" measurement. In the technolgies I work with, the delay matrix is 7x7, but setup / hold timing has only 3x3. It would be good to split the matrix for IO delays (`input slew` , `output cap`) from the setup / hold matrix (`constrained pin slew` , `related pin slew`). Now the `slew` is taken for both which is confusing. E.g. from the liberty file:
```
 /* Table Templates */
  lu_table_template (delay_template_3x3) {
    variable_1 : input_net_transition;
    variable_2 : total_output_net_capacitance;
    index_1 ("0.002, 0.162966, 0.8");
    index_2 ("0.001, 0.00540346, 0.0228305");
  }
  lu_table_template (setup_template_3x3) {
    variable_1 : related_pin_transition;
    variable_2 : constrained_pin_transition;
    index_1 ("0.002, 0.162966, 0.8");
    index_2 ("0.001, 0.00540346, 0.0228305");
  }
```
2. When I try to characterize a flip-flop, I get vastly different results compared to FAB provided data
at the same operating conditions (TT, 25 C, 1.2V):

E.g. charlib result for D pin setup:
```
rise_constraint (setup_template_3x3) {
          index_1 ("0.002, 0.162966, 0.8");
          index_2 ("0.001, 0.00540346, 0.0228305");
          values ( \
            "1.402000, 1.402000, 1.402000", \
            "1.402000, 1.402000, 1.402000", \
            "1.402000, 1.402000, 1.402000");
        }
```
While if I look at the setup time of D pin for (0.002, 0.002) from foundry provided `.lib`
file, I get sth like 0.046. The difference is something like factor of 35.

Further, all the setup times are the same, which is not the same in the foundry provided
data.

I do not have SPICE netlist extracted with parasitics from layout of the cell.
I only have "LVS" netlist -> List of transistors and their connections.
Maybe since the measurement occurs at range where circuit is metastable,
the output delay may significantly vary. But I don't know.

Do you have any clue on what may have went wrong ?

**Expected behavior**
Be within some reasonable range from the FAB characterized `.lib` file (e.g. +-20%)

### PR DESCRIPTION
**Pull request by AndersonHsieh0330**
initial implementation of the setup and hold search, waiting for review with Marcus

### CODE REVIEW

File: charlib/characterizer/procedures/combinational_delay.py [hunk 1]
```diff
@@ -180,7 +180,8 @@ def measure_delay_for_path_with_criterion(cell, config, settings, variation, pat
             plt.close(fig)
 
         # Build LUT
-        delay = criterion([analysis.measurements[name] for analysis in analyses.values()]) @ PySpice.Unit.u_s
+        delay = criterion([analysis.measurements[name] for analysis in analyses.values()
+                           if name in analysis.measurements]) @ PySpice.Unit.u_s
```
Comment (infinitymdm) [inline charlib/characterizer/procedures/combinational_delay.py]: Small style nitpick: it might be better to split this comprehension into two lines e.g.
```python
delay_measurements = [analysis.measurements[name] ...]
delay = criterion(delay_measurements @ PySpice.Unit.u_s)
```
It may even be best to define a generator to get the measurements. That could probably go into `utils` or `procedures`, as this won't be the only place this is used.

File: charlib/characterizer/cell.py [hunk 2]
```diff
@@ -243,9 +243,19 @@ def nonmasking_conditions_for_path(self, input_port, input_transition, output_po
...
-                tv[output_port] = tv.pop(Function.OUT)
-                yield tv
+            pin_val = tv[input_port]
+
+            # Combinational: pin_val is '01' or '10' (the pin is transitioning in the vector)
+            # Sequential: pin_val is '0' or '1' (the pin is static because the clock transitions)
+            if len(pin_val) == 2:
+                input_match = (pin_val == input_transition)
+            else:
+                input_match = (pin_val == input_transition[-1])
```
Comment (infinitymdm) [inline charlib/characterizer/cell.py:253]: This could be replaced by `(tv[input_port] == list(input_transition)[-1])`

File: charlib/characterizer/cell.py [hunk 3]
```diff
@@ -243,9 +243,19 @@ def nonmasking_conditions_for_path(self, input_port, input_transition, output_po
...
+
+            if input_match and tv[Function.OUT] == output_transition:
+                result = {**tv}
+                result[output_port] = result.pop(Function.OUT)
```
Comment (infinitymdm) [inline charlib/characterizer/cell.py]: What's the purpose of these two lines? Isn't `result` just the same thing as `tv`?

File: charlib/characterizer/plots.py
Comment (infinitymdm) [inline charlib/characterizer/plots.py:1]: I like that you split this out. Nice improvement.

File: charlib/config/syntax.py [hunk 4]
```diff
@@ -139,6 +148,30 @@ class ConfigFile:
...
+                            'setting in ``settings.simulation.sequential_setup_hold_procedure``.'
+            ), default='sequential_setup_hold'
+        ) : str,
+        Optional(
+            Literal(
+                'sequential_c_load',
+                description='Capacitive load applied to the output during sequential ' \
+                            'characterization. Unit is specified by ``settings.units.capacitive_load``. ' \
+                            'Defaults to 0.1.'
+            ), default=0.1
```
Comment (infinitymdm) [inline charlib/config/syntax.py]: Convince me that this is a better key to use instead of reusing the existing `loads` key (and keeping it required for all cells).

File: charlib/config/syntax.py [hunk 5]
```diff
@@ -60,10 +60,12 @@ class ConfigFile:
             description='A list of input pin slew rates to characterize. Unit is specified by ' \
                         '``settings.units.time``.'
         ) : [Or(float, int)],
-        Literal(
-            'loads',
-            description='A list of output capacitive loads to characterize. Unit is specified ' \
-                        'by ``settings.units.capacitive_load``.'
+        Optional(
```
Comment (infinitymdm) [inline charlib/config/syntax.py:63]: See comment on `sequential_c_load`: https://github.com/stineje/CharLib/pull/88/changes#r2925675286

File: charlib/config/syntax.py [hunk 6]
```diff
@@ -103,6 +105,13 @@ class ConfigFile:
                             'asynchronous reset pin).'
             )
         ) : Regex('^(posedge|negedge|not|!|())[ ]*[a-zA-Z0-9_]+'),
+        Optional(
+            Literal(
+                'flops',
+                description='A list of internal node names used as state storage elements in a ' \
+                            'sequential cell (e.g. internal latch nodes).'
+            )
+        ) : [str],
```
Comment (infinitymdm) [inline charlib/config/syntax.py]: Internal latch names should be derived from `state` and/or `functions` instead. See https://github.com/stineje/CharLib/issues/64#issuecomment-3423842730 for details on usage.

File: charlib/config/syntax.py [hunk 7]
```diff
@@ -139,6 +148,30 @@ class ConfigFile:
...
             )
         ) : [Or(float, int)],
+        Optional(
+            Literal(
+                'sequential_setup_hold_procedure',
+                description='The name of a procedure used to measure setup and hold time ' \
+                            'constraints for a sequential cell. Overrides the library-wide ' \
+                            'setting in ``settings.simulation.sequential_setup_hold_procedure``.'
+            ), default='sequential_setup_hold'
+        ) : str,
```
Comment (infinitymdm) [inline charlib/config/syntax.py]: This is interesting. What's the case to be made for allowing per-cell procedure selection? Shouldn't the whole library use the same procedures for a particular measurement (for consistency)?

File: charlib/characterizer/procedures/sequential_delay.py [hunk 8]
```diff
@@ -20,19 +27,586 @@ def measure_removal_constraint(cell_settings, charlib_settings, control_pin, tri
...
 
-def binary_search_setup_hold_constraint(cell_settings, charlib_settings, data_pin, trigger_pin, state_pin):
-    """Find the minimum time the data pin must be set preceding and following the trigger.
+@register
+def sequential_delay(cell_settings, charlib_settings, data_pin, trigger_pin, state_pin):
+    """Measure the delay between trigger activation and state change.
+
+    For clock-edge-triggered devices, this is commonly called the clock-to-Q or C2Q delay. This
+    value depends on the transition time of both the data pin and the trigger pin."""
+    pass # TODO
```
Comment (infinitymdm) [inline charlib/characterizer/procedures/sequential_delay.py]: Remove - your procedure replaces this.

File: charlib/characterizer/procedures/sequential_delay.py [hunk 9]
```diff
@@ -20,19 +27,586 @@ def measure_removal_constraint(cell_settings, charlib_settings, control_pin, tri
...
+        cs = variation['clock_slew'] * charlib_settings.units.time
+        variation_debug_path = charlib_settings.debug_dir / cell.name / __name__.split('.')[-1] / f'variation_data_slew_{str(ds).replace(' ', '')}_clock_slew_{str(cs).replace(' ', '')}'
+        variation_debug_path.mkdir(parents=True, exist_ok=True)
+
+        for path in cell.paths():
+            # cell.nonmasking_conditions_for_path filter out the impossible paths
+            # ex. non-inverting FF with D, Q, and CLK. it'll never have D_01 -> Q_10
+            state_maps = list(cell.nonmasking_conditions_for_path(*path))
+            if state_maps == []:
+                continue
```
Comment (infinitymdm) [inline charlib/characterizer/procedures/sequential_delay.py:51]: I'm pretty sure you should be able to remove this check and move the state mapping into `find_setup_hold_simple_for_variation`. StateFunction should now filter out impossible paths during TV generation (though there are some issues with inverting outputs at present).

File: charlib/characterizer/procedures/sequential_delay.py [hunk 10]
```diff
@@ -20,19 +27,586 @@ def measure_removal_constraint(cell_settings, charlib_settings, control_pin, tri
...
-    For clock-edge-triggered devices, this is commonly called the clock-to-Q or C2Q delay. This
-    value depends on the transition time of both the data pin and the trigger pin."""
+    TOLERANCE = 10 @ PySpice.Unit.u_ps # binary search stops when iternation n - (n-1) < TOLERANCE
+    STEP = 5 @ PySpice.Unit.u_ps # 5 ps initial bound-finding step for binary search
+    T_STABILZING = 20 * max(cell_settings.parameters['clock_slews']) * charlib_settings.units.time # 20x worst case clock slew
+    # C_LOAD = 4x the data pin's input capacitance from the prior ac_sweep, (fanout of 4)
+    # C_LOAD = 4 * cell.liberty.group('pin', data_pin).attributes['capacitance'].value * settings.units.capacitance
+    C_LOAD = cell_settings.parameters['sequential_c_load'] * charlib_settings.units.capacitance
+    N_SWEEP_SAMPLES = cell_settings.parameters['sequential_n_sweep_samples']
+    constants = (TOLERANCE, STEP, T_STABILZING, C_LOAD, N_SWEEP_SAMPLES)
```
Comment (infinitymdm) [inline charlib/characterizer/procedures/sequential_delay.py]: I'd prefer these constants be generated within the `find_setup_hold_simple_for_variation` procedure rather than this generator. Same goes for `variation_debug_path`.

File: charlib/characterizer/procedures/sequential_delay.py [hunk 11]
```diff
@@ -20,19 +27,586 @@ def measure_removal_constraint(cell_settings, charlib_settings, control_pin, tri
...
+        write_step_log(step4_path, 'step4', cell.name, ds, cs, path_str, state_str, constants,
+                        f"step4 : find setup given hold= {step3_hold_result}",
+                        step4_setup_result, step4_phase1_candidates, step4_phase2_candidates)
+        
+        # Step 5: sweep the setup×hold boundary and plot the latched contour
+        step5_debug_path = (state_debug_path / 'step5') if settings.debug else None
+        ref_c2q = get_c2q(cell, config, settings, cs, ds,
+                          step4_setup_result, step2_hold_result,
+                          C_LOAD, T_STABILZING, path, state_map, step5_debug_path)
+        c2q_threshold = ref_c2q * 1.2 if not math.isnan(ref_c2q) else math.inf
```
Comment (infinitymdm) [inline charlib/characterizer/procedures/sequential_delay.py:201]: What do we think: should the "back-off factor" of 1.2 here be customizable from settings?

File: charlib/characterizer/procedures/sequential_delay.py [hunk 12]
```diff
@@ -20,19 +27,586 @@ def measure_removal_constraint(cell_settings, charlib_settings, control_pin, tri
...
-    both constraints."""
+    For example, for a rising-edge DFF with an asynchronous reset, the minimum pulse width is the
+    minimum time the clock signal must be active before the rising edge in order to reset the
+    device state."""
     pass # TODO
 
-def measure_delays(cell_settings, charlib_settings, data_pin, trigger_pin, state_pin):
-    """Measure the delay between trigger activation and state change.
+@register
+def sequential_setup_hold(cell, cell_settings, charlib_settings):
```
Comment (infinitymdm) [inline charlib/characterizer/procedures/sequential_delay.py]: Indicate methodology in the procedure generator name. I'd suggest something like `measure_setup_hold_from_contour`

File: charlib/characterizer/utils.py
Comment (infinitymdm) [inline charlib/characterizer/utils.py:1]: Some of the new methods here are pretty specific to setup & hold search. I think we may need to put some thought into how we organize utils. (This is more of a note for future me than a review comment)

**Review by infinitymdm — CHANGES_REQUESTED** [review]
Needs a few minor changes and a little bit of cleanup, but overall this looks like a big improvement compared to prior sequential procedures. 

Once you address the requested changes, I'd like to test this with gf180mcu as well.

File: charlib/characterizer/cell.py (hunk 3 above)
Comment (infinitymdm) [reply charlib/characterizer/cell.py]: Ah, I see what you're doing. This replaces the generic output key (since the function doesn't know what output it targets) with the actual port name. You should be able to make this more concise: `tv[output_port] = tv.pop(Function.OUT)`. Then you don't have to duplicate the whole dict.

**Comment by infinitymdm** [comment]
Note re. failing gf180mcu runs: our gf180mcu_osu testing config.yaml needs to be updated to work with the new config syntax introduced in this PR.

File: charlib/config/syntax.py (hunk 7 above)
Comment (AndersonHsieh0330) [reply charlib/config/syntax.py]: yes initially I was going to implement two procedures, one simple one (1 binary search for setup, one for hold and that's it), one complex one (what I have here, implements what the paper describes). I didn't end up creating a simple one cuz it didn't make sense so I just the procedure field for YAML in case we do want to add other procedures for sequential setup / hold search in the future. I can remove this

File: charlib/config/syntax.py (hunk 4 above)
Comment (AndersonHsieh0330) [reply charlib/config/syntax.py]: "loads" should take a list and it is usually used for the delay table (capacitive load vs input data slew). This "sequential_c_load" is one capacitive load value used across every single simulation for setup and hold, across very clk slew vs data slew variation. I didn't want to confuse the user thus this was created.

the "loads" key which takes a list of loads would still be used for C2Q delay table. such table will be written out to the timing section that belongs to pin(Q)

File: charlib/characterizer/procedures/combinational_delay.py (hunk 1 above)
Comment (AndersonHsieh0330) [reply charlib/characterizer/procedures/combinational_delay.py]: So I added the if statement check since it was failing for some cases, and I couldn't remember which cell it was exactly. This means "analysis.measurements" sometimes may not have all the measurement entries that we want it to take. 

In other words, Pyspice attempts to take all the measurements listed in "measurements (local list)" but when it gets returned, some measurements are not registered into "analysis.measurements"

I think this suggests that nonmasking_conditions_for_path is not filtering out all the cases where the output pin does not change. Did I get that correctly?

If so, I can modify it into the two line format that you specified.

File: charlib/characterizer/cell.py (hunk 3 above)
Comment (AndersonHsieh0330) [reply charlib/characterizer/cell.py]: agreed, changes incoming

File: charlib/characterizer/cell.py (hunk 2 above)
Comment (AndersonHsieh0330) [reply charlib/characterizer/cell.py:253]: agreed, changes incoming

File: charlib/characterizer/procedures/combinational_delay.py (hunk 1 above)
Comment (infinitymdm) [reply charlib/characterizer/procedures/combinational_delay.py]: I have no issue with the addition of the filter for making sure all the measurement names are there. This is purely a code style thing - I'm trying to avoid splitting a list comprehension (in the midst of another expression) across multiple lines.

File: charlib/config/syntax.py (hunk 4 above)
Comment (infinitymdm) [reply charlib/config/syntax.py]: Fair enough; I'm convinced that we need this key. Can we workshop the name a bit though? `sequential_c_load` (and the description, for that matter) makes this seem like it applies to all sequential characterization tasks, which is not the case if I understand correctly.

Off the top of my head I'm thinking `setup_hold_constraint_load` or `metastability_constraint_load`. Feel free to come up with something better.

File: charlib/config/syntax.py (hunk 5 above)
Comment (infinitymdm) [reply charlib/config/syntax.py:63]: I think sequential and combinational cells will both still need the `loads` key. If you can think of a case where this key is not needed, making it optional is fine. Otherwise this `loads` should remain a required key for cell entries.

File: charlib/config/syntax.py (hunk 7 above)
Comment (infinitymdm) [reply charlib/config/syntax.py]: I actually don't totally disagree with the idea here. There may be something to be said for allowing different simulation parameters (such as procedure) to be specified on a per-cell basis. For now though, probably best to remove; that's something we can explore in the future.

File: charlib/characterizer/procedures/sequential_delay.py (hunk 9 above)
Comment (infinitymdm) [reply charlib/characterizer/procedures/sequential_delay.py:51]: Feel free to ignore this if your testing indicates that there are still significant enough issues with inverting paths

File: charlib/characterizer/procedures/sequential_delay.py (hunk 11 above)
Comment (stineje) [reply charlib/characterizer/procedures/sequential_delay.py:201]: I think this is a good idea although most units use the term back-off distance or delay growth threshold.  I have seen guardband or push-out method too.  Check on this paper @infinitymdm https://ieeexplore.ieee.org/document/4195625

Comment (infinitymdm) [reply charlib/characterizer/procedures/sequential_delay.py:201]: `delay_growth_threshold` (or maybe `c2q_backoff_threshold`?) would be a good name for this key in the yaml. The 20% back-off here is the number used in Salman et. al's paper; that gives us a good default. 

@AndersonHsieh0330 Feel free to leave this as-is. We'll open a new issue once this PR is merged to make this value configurable from the YAML.

**Comment by infinitymdm** [comment]
In my testing with gf180mcu_osu, I'm seeing step 0 of the `find_setup_hold_simple_for_variation` function fail to return a valid time. Still looking into why. I'll get back to you on that once I have more info.

File: charlib/characterizer/procedures/combinational_delay.py (hunk 1 above)
Comment (AndersonHsieh0330) [reply charlib/characterizer/procedures/combinational_delay.py]: gotcha, changes in coming

File: charlib/config/syntax.py (hunk 4 above)
Comment (AndersonHsieh0330) [reply charlib/config/syntax.py]: yes i like `setup_hold_constraint_load`, changes in coming

File: charlib/characterizer/cell.py (hunk 2 above)
Comment (AndersonHsieh0330) [reply charlib/characterizer/cell.py:253]: wait sorry I thought about it more, if I replace that whole block by 
```
input_match = (tv[input_port] == list(input_transition)[-1])
```
Then in the case of 
tv[input_port]=01, input_transition=01, then it'll always return false.

Am I missing something?

File: charlib/config/syntax.py (hunk 5 above)
Comment (AndersonHsieh0330) [reply charlib/config/syntax.py:63]: i think we'll use "loads" for C2Q delay for sequential cells, I'll leave it as optional for now and once c2q time characterization is implemented will move it back to required. is that okay?

File: charlib/config/syntax.py (hunk 6 above)
Comment (AndersonHsieh0330) [reply charlib/config/syntax.py]: yeah that's my bad, didn't realize internal latch names are inferred by state now. I'll remove the flops keyword

File: charlib/characterizer/cell.py (hunk 3 above)
Comment (AndersonHsieh0330) [reply charlib/characterizer/cell.py]: did more testing, we have to duplicate the dictionary.  Function.test_vectors caches its result in self.stored_test_vectors. removing the duplicated code called tv.pop(Function.OUT) directly on the cached dict, which mutated it in-place and removed the __output key.
On the next calls, the cached vectors no longer had __output, causing a KeyError. 

we can either replace the output pin name in self.stored_test_vector as soon as we initialize the cell, so we don't have to query output transition by Function.OUT, or we duplicate the dictionary here

File: charlib/characterizer/procedures/sequential_delay.py (hunk 9 above)
Comment (AndersonHsieh0330) [reply charlib/characterizer/procedures/sequential_delay.py:51]: I'm going to keep this because there is parallelism benefit — each (variation, path) combination is
  yielded as a separate task and submitted to ProcessPoolExecutor independently, so simulations from multiple paths can be ran in parallel.

Comment (AndersonHsieh0330) [reply charlib/characterizer/procedures/sequential_delay.py:51]: the naming of the function "find_setup_hold_for_variation" will be changed

File: charlib/characterizer/cell.py (hunk 2 above)
Comment (infinitymdm) [reply charlib/characterizer/cell.py:253]: No, you're right. My mistake there. This should be good.

File: charlib/characterizer/procedures/sequential_delay.py (hunk 10 above)
Comment (AndersonHsieh0330) [reply charlib/characterizer/procedures/sequential_delay.py]: sounds good, changes incoming

File: charlib/characterizer/procedures/sequential_delay.py (hunk 11 above)
Comment (AndersonHsieh0330) [reply charlib/characterizer/procedures/sequential_delay.py:201]: sounds good, will leave as it is

File: charlib/characterizer/procedures/sequential_delay.py (hunk 12 above)
Comment (AndersonHsieh0330) [reply charlib/characterizer/procedures/sequential_delay.py]: renamed in new commit
good name

**Review by infinitymdm — APPROVED** [review]
LGTM. gf180mcu_osu_sc runs in CI need modifications and can be ignored for now. Let's get this merged.
